### ISSUE TITLE
Any plans on supporting protobuf 3?

### ISSUE THREAD

**Issue by Pacane** [dart-lang/dart-protoc-plugin#59]
^

**Comment by skybrian** [comment]
No concrete plans. (Certainly nice to have, but nobody is working on it.)

**Comment by skybrian** [comment]
An intermediate step might be to make sure that we can interop with proto 3. (Apparently it should be forward/backward compatible.)

**Comment by Pacane** [comment]
Well that's not the case right now. I tried compiling some proto3 stuff with this, but nope!

**Comment by Pacane** [comment]
Also, [this](https://github.com/dart-lang/dart-protoc-plugin/issues/58) is going to be a blocker to update anything in this lib I believe.

**Comment by nemequ** [comment]
It's worth noting that this blocks proper [gRPC](http://www.grpc.io/) support.

**Comment by Brahmasmi** [comment]
Would it be possible to please share the current status of support for Protobuf 3?

The intended use case is to use grpc-dart in a flutter app for the communication stack. This would help replace Java and Swift grpc communication stacks with a common Dart one.

Thanks.

@jakobr-google @mit-mit @skybrian

**Comment by jakobr-google** [comment]
Sure! The current status is that proto3 support mostly works, since the wire format is backwards compatible, but we have a bit of work to do before we can claim full support. 

So, gRPC services defined in a proto3 document should work, but with some caveats:
* The JSON representation doesn't match the proto3 spec. There's work in progress to fix this.
* Proto3 fields don't have special support, specifically:
  * Maps show up a lists of entries instead of a proper map in Dart
  * OneOf fields don't properly clear the other fields when a new one is set and we don't yet have an accessor to show which field is set
  * Any fields will probably just show up as a blob.

Apart from that, compiling a proto3 definition to Dart code should work, including gRPC services. If you're not using any of the above, you should be fine. If you are, you'll need to work around the missing support for now.

Full proto3 support is coming up.

**Comment by Brahmasmi** [comment]
Danke @jakobr-google for the detailed update. It feels great to see the possibility of a common gRPC Dart communication stack.

Would it be possible for you to share an approximate timeline for the full proto3 support? Q1/2/3/4 2018?

Thanks.

**Comment by sigurdm** [comment]
Any fields are now supported with a pack and unpack method!

**Comment by sigurdm** [comment]
> > The JSON representation doesn't match the proto3 spec. There's work in progress to fix this.
> 
> Any progress on this?

This is being worked on here: https://github.com/dart-lang/protobuf/pull/83
We are going to support both json with name-keys and with numeric keys

**Comment by Brahmasmi** [comment]
Tak Sigurd for the Any support.

Looking forward to the Map and OneOf i.e. complete Protobuf 3 support.

Dhanyavaad.
